src/common/io.cc

```diff
@@ -6,14 +6,96 @@
 #include <fcntl.h>
 #include <unistd.h>
 #endif  // defined(__unix__)
+#include <algorithm>
 #include <cstdio>
 #include <string>
 
 #include "xgboost/logging.h"
+#include "io.h"
 
 namespace xgboost {
 namespace common {
 
+size_t PeekableInStream::Read(void* dptr, size_t size) {
+  size_t nbuffer = buffer_.length() - buffer_ptr_;
+  if (nbuffer == 0) return strm_->Read(dptr, size);
+  if (nbuffer < size) {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, nbuffer);
+    buffer_ptr_ += nbuffer;
+    return nbuffer + strm_->Read(reinterpret_cast<char*>(dptr) + nbuffer,
+                                 size - nbuffer);
+  } else {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, size);
+    buffer_ptr_ += size;
+    return size;
+  }
+}
+
+size_t PeekableInStream::PeekRead(void* dptr, size_t size) {
+  size_t nbuffer = buffer_.length() - buffer_ptr_;
+  if (nbuffer < size) {
+    buffer_ = buffer_.substr(buffer_ptr_, buffer_.length());
+    buffer_ptr_ = 0;
+    buffer_.resize(size);
+    size_t nadd = strm_->Read(dmlc::BeginPtr(buffer_) + nbuffer, size - nbuffer);
+    buffer_.resize(nbuffer + nadd);
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_), buffer_.length());
+    return buffer_.length();
+  } else {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, size);
+    return size;
+  }
+}
+
+FixedSizeStream::FixedSizeStream(PeekableInStream* stream) : PeekableInStream(stream), pointer_{0} {
+  size_t size {1}, total {0};
+  buffer_.clear();
+  while (true) {
+    buffer_.resize(size);
+    size_t read = stream->PeekRead(&buffer_[0], size);
+    total += read;
+    if (read < size) {
+      break;
+    }
+    size *= 2;
+  }
+  buffer_.resize(total);
+}
+
+size_t FixedSizeStream::Read(void* dptr, size_t size) {
+  auto read = this->PeekRead(dptr, size);
+  pointer_ += read;
+  LOG(WARNING) << "read:" << read << "\n"
+               << "Buffer:";
+  for (size_t i = pointer_; i < buffer_.size(); ++i) {
+    std::cout << buffer_[i];
+  }
+  std::cout << std::endl;
+  return read;
+}
+
+// size_t FixedSizeStream::Read(std::string* str) {
+//   return this->Read(&((*str)[0]), str->size());
+// }
+
+size_t FixedSizeStream::PeekRead(void* dptr, size_t size) {
+  if (size >= buffer_.size() - pointer_)  {
+    LOG(WARNING) << "Read exceeded";
```

Comment:
Why do we have a warning here but not line 37? Should one write logs in both cases for consistency?

Also I am not sure the line "Read exceeded" will be that useful without any additional information

Reply:
@trams Nah, that's just me debugging.  Will not appear in final code.  It's an WIP.  @RAMitchell suggested to split this PR into smaller portions.  Will try doing that once I can get everything sorted our and document the whole logic clearly.